lib/utils/fanoutbuffer/buffer.go

```diff
@@ -0,0 +1,406 @@
+	"github.com/jonboulle/clockwork"
+)
+
+// GracePeriodExceeded is an error returned by Cursor.Read indicating that the cursor fell
+// too far behind. Observing this error indicates that either the reader is too slow, or the
+// buffer has been configured with insufficient capacity/backlog for the usecase.
+var GracePeriodExceeded = errors.New("failed to process fanout buffer backlog within grace period")
+
+// UseOfClosedCursor is an error indicating that Cursor.Read was called after the cursor had
+// either been explicitly closed, or had previously returned an error.
+var UseOfClosedCursor = errors.New("use of closed fanout buffer cursor (this is a bug)")
+
+// BufferClosed is an error indicating that the event buffer as a whole has been closed.
+var BufferClosed = errors.New("fanout buffer closed")
+
+// Config holds all configuration parameters for the fanout buffer. All parameters are optional.
+type Config struct {
+	// Capacity is the capacity to allocate for the main circular buffer. Capacity should be selected s.t. cursors rarely
+	// fall behind capacity during normal operation.
+	Capacity uint64
+
+	// GracePeriod is the amount of time a backlog (beyond the specified capacity) will be allowed to persist. Longer grace
+	// periods give cursors more time to catch up in the event of spikes, at the cost of higher potential memory usage and
+	// longer waits before unhealthy cursors are ejected.
+	GracePeriod time.Duration
+
+	// Clock is used to override default time-behaviors in tests.
+	Clock clockwork.Clock
+}
+
+// SetDefaults sets default config parameters.
+func (c *Config) SetDefaults() {
+	if c.Capacity == 0 {
+		c.Capacity = 512
+	}
+
+	if c.GracePeriod == 0 {
+		c.GracePeriod = time.Second * 30
+	}
+
+	if c.Clock == nil {
+		c.Clock = clockwork.NewRealClock()
+	}
+}
+
+// Buffer is a circular buffer that keeps track of how many cursors exist, and how many have seen
+// each item, so that it knows when items can be cleared. If one or more cursors fall behind, the items
+// they have yet to see go into a temporary backlog of infinite size. If the backlog persists for greater
+// than the allowed grace period, it is cleared and all cursors still in the backlog fail on their next
+// attempt to read.
+type Buffer[T any] struct {
+	rw sync.RWMutex
+
+	// cfg holds buffer configuration.
+	cfg Config
+
+	// notify is closed and replaced each time an item is added to wake any cursors
+	// that are fully caught up.
+	notify chan struct{}
+
+	// closed indicates that the buffer has been closed.
+	closed bool
+
+	// cursors is the total number of cursors that exist. used to determine the number of cursors expected
+	// to observe each item.
+	cursors uint64
+
+	// pos is the absolute position of the oldest elemet in the ring, meaning that pos%Capacity is the index
+	// of that element.
+	pos uint64
+
+	// len is the total number of elements in the ring (i.e. the index of the newest element is at `(pos + len - 1) % Capacity`).
+	len uint64
```

Comment:
I'm not sure the `uin64` vs `uint` matters.  Go's native indexing/length repr is `int`, anything else means a cast.  I chose `uint64` for `pos` because it's monotonically increasing.  Once one of these values is `uint64`, it becomes simpler to just have all of them be `uint64` I think.